wire-schema/src/main/java/com/squareup/wire/schema/internal/parser/OptionElement.kt

```diff
@@ -0,0 +1,115 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.squareup.wire.schema.internal.parser
+
+import com.squareup.wire.schema.internal.Util
+import com.squareup.wire.schema.internal.Util.appendIndented
+
+data class OptionElement(
+  val name: String,
+  val kind: Kind,
+  val value: Any,
+  private val isParenthesized: Boolean
+) {
+  enum class Kind {
+    STRING,
+    BOOLEAN,
+    NUMBER,
+    ENUM,
+    MAP,
+    LIST,
+    OPTION
+  }
+
+  fun toSchema(): String = buildString {
+    when (kind) {
+      Kind.STRING -> return """${formatName()} = "$value""""
+      Kind.BOOLEAN,
+      Kind.NUMBER,
+      Kind.ENUM -> return "${formatName()} = $value"
+      Kind.OPTION -> {
+        // Treat nested options as non-parenthesized always, prevents double parentheses.
+        val optionValue = (value as OptionElement).run {
+          OptionElement.create(name, kind, value)
+        }
+        append("${formatName()}.${optionValue.toSchema()}")
+      }
+      Kind.MAP -> {
+        append("${formatName()} = {\n")
+        formatOptionMap(this, value as Map<String, *>)
+        append('}')
+      }
+      Kind.LIST -> {
+        append("${formatName()} = ")
+        Util.appendOptions(this, value as List<OptionElement>)
+      }
+    }
+  }
+
+  fun toSchemaDeclaration() = "option ${toSchema()};\n"
+
+  private fun formatName() = if (isParenthesized) "($name)" else name
+
+  companion object {
+    @JvmOverloads
+    fun create(
+      name: String,
+      kind: Kind,
+      value: Any,
+      isParenthesized: Boolean = false
+    ) = OptionElement(name, kind, value, isParenthesized)
+
+    private fun formatOptionMap(
+      builder: StringBuilder,
+      valueMap: Map<String, *>
+    ) {
+      valueMap.entries.forEach { entry ->
+        val endl = if (entry != valueMap.entries.last()) "," else ""
+        appendIndented(
+            builder,
+            "${entry.key}: ${formatOptionMapValue(entry.value!!)}$endl"
+        )
+      }
+    }
+
+    private fun formatOptionMapValue(value: Any): String = buildString {
+      when (value) {
+        is String -> {
+          return """"$value""""
+        }
+        is Map<*, *> -> {
+          append("{\n")
+          formatOptionMap(this, value as Map<String, *>)
+          append('}')
+        }
+        is List<*> -> {
+          append("[\n")
+          var i = 0
+          val count = value.size
+          while (i < count) {
+            val endl = if (i < count - 1) "," else ""
+            appendIndented(this, formatOptionMapValue(value[i]!!) + endl)
+            i++
+          }
+          append("]")
+        }
+        else -> {
+          return value.toString()
```

Comment:
nit: `append(value)` (and for `String`). otherwise move the `buildString` into the `when` arms. Mixing `buildString` and `return` throws me off... plus you've already allocated the `StringBuilder` at this point, might as well use it.